gno.land/pkg/gnoweb/url.go

```diff
@@ -28,121 +32,147 @@ type GnoURL struct {
+func (gnoURL GnoURL) EncodeArgs() string {
+	return gnoURL.Encode(EncodeArgs | EncodeQuery | EncodeNoEscape)
+}
+
+// EncodeURL encodes the path, arguments, and query parameters into a string.
+// This function provides the full representation of the URL without the web query.
+func (gnoURL GnoURL) EncodeURL() string {
+	return gnoURL.Encode(EncodePath | EncodeArgs | EncodeQuery)
+}
+
+// EncodeWebURL encodes the path, package arguments, web query, and query into a string.
+// This function provides the full representation of the URL.
+func (gnoURL GnoURL) EncodeWebURL() string {
+	return gnoURL.Encode(EncodePath | EncodeArgs | EncodeWebQuery | EncodeQuery)
+}
+
+// Kind determines the kind of path (invalid, realm, or pure) based on the path structure.
+func (gnoURL GnoURL) Kind() PathKind {
+	if len(gnoURL.Path) > 2 && gnoURL.Path[0] == '/' && gnoURL.Path[2] == '/' {
+		switch k := PathKind(gnoURL.Path[1]); k {
+		case KindPure, KindRealm:
+			return k
+		}
 	}
-	return KindInvalid
+	return KindUnknown
 }
 
-var (
-	ErrURLMalformedPath   = errors.New("malformed URL path")
-	ErrURLInvalidPathKind = errors.New("invalid path kind")
-)
+func (gnoURL GnoURL) IsValid() bool {
+	return rePkgOrRealmPath.MatchString(gnoURL.Path)
+}
 
-// reRealName match a realm path
-// - matches[1]: path
-// - matches[2]: path args
-var reRealmPath = regexp.MustCompile(`^` +
-	`(/(?:[a-zA-Z0-9_-]+)/` + // path kind
-	`[a-zA-Z][a-zA-Z0-9_-]*` + // First path segment
-	`(?:/[a-zA-Z][.a-zA-Z0-9_-]*)*/?)` + // Additional path segments
-	`([:$](?:.*))?$`, // Remaining portions args, separate by `$` or `:`
-)
+// IsDir checks if the URL path represents a directory.
+func (gnoURL GnoURL) IsDir() bool {
+	return len(gnoURL.Path) > 0 && gnoURL.Path[len(gnoURL.Path)-1] == '/'
+}
 
+// IsFile checks if the URL path represents a file.
+func (gnoURL GnoURL) IsFile() bool {
+	return filepath.Ext(gnoURL.Path) != ""
+}
+
+var ErrURLInvalidPath = errors.New("invalid or malformed path")
+
+// ParseGnoURL parses a URL into a GnoURL structure, extracting and validating its components.
 func ParseGnoURL(u *url.URL) (*GnoURL, error) {
-	matches := reRealmPath.FindStringSubmatch(u.EscapedPath())
-	if len(matches) != 3 {
-		return nil, fmt.Errorf("%w: %s", ErrURLMalformedPath, u.Path)
+	var webargs string
+	path, args, found := strings.Cut(u.EscapedPath(), ":")
+	if found {
+		args, webargs, _ = strings.Cut(args, "$")
+	} else {
+		path, webargs, _ = strings.Cut(path, "$")
 	}
 
-	path := matches[1]
-	args := matches[2]
+	// NOTE: `PathUnescape` should already unescape dollar signs.
+	upath, err := url.PathUnescape(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unescape path %q: %w", path, err)
+	}
 
-	if len(args) > 0 {
-		switch args[0] {
-		case ':':
-			args = args[1:]
-		case '$':
-		default:
-			return nil, fmt.Errorf("%w: %s", ErrURLMalformedPath, u.Path)
-		}
+	if !rePkgOrRealmPath.MatchString(upath) {
```

Comment:
What if we used `gnolang.IsPurePackagePath` / `gnolang.IsRealmPath` instead?

Reply:
To check the string path directly? Or do you mean removing the check and using `IsPurePackagePath` and `IsRealmPath`? I would like to have `IsPurePackagePath` and `IsRealmPath`; I think it's a better idea than `Kind()`, but I'm not sure why you pointed out this specific line.

Reply:
I just had beef with this regex and this seemed a good place where to make a comment :D

Reply:
But yeah, I think we can also substitute this regex to determine if it's a valid path?

Reply:
I don't think we should forbid a URL that is neither pure nor a realm. I prefer this package to perform minimal checks; it should be the caller's choice to accept the URL based on the parsed information. That said, we can still establish a minimal base to work on without assuming _any_ urls.

What do you think about:
-  Removing the concept of "kind" within the structure while still providing helpers to check if the method is pure or a realm, as you mentioned.
-  Refining the regex to define what a path can be, based on what we have in [gnovm/pkg/gnolang/helpers.go](https://github.com/gnolang/gno/blob/88fb8cb3aac1e3b393a9f8e88304b78fbe1240da/gnovm/pkg/gnolang/helpers.go#L15-L18), with the only base constraint that it must start with a slash and at least one character. This will makes the check and the assumption of the path structure simple and easy to manage.
```
var rePkgOrRealmPath = regexp.MustCompile(`^/[a-z][a-z0-9_/.]*$`)
```

I understand that currently only pure and realm exist, but `gnoweb` will likely extend this with `/u`, for example. So it makes sense to me to make `GnoURL` a bit more flexible in what it can parse. This way, the caller can decide whether to use it as a direct call to the node or to redirect the user to something else.

Reply:
@thehowl suggestion preview: https://github.com/gfanton/gno/commit/d0eb4b098183313634eecd29b30336e4e6197e7f